File: services/menos/menos/services/youtube_metadata.py
"""YouTube metadata fetching service using Data API v3."""

import re
from dataclasses import dataclass
from datetime import datetime
from typing import Any

from menos.config import settings


@dataclass
class YouTubeChannelVideo:
    """YouTube channel video list item."""

    video_id: str
    title: str
    url: str
    published_at: str
    duration: str | None = None
    duration_seconds: int | None = None
    view_count: int | None = None

    def to_dict(self) -> dict:
        """Convert to dictionary for JSON serialization."""
        return {
            "video_id": self.video_id,
            "title": self.title,
            "url": self.url,
            "published_at": self.published_at,
            "duration": self.duration,
            "duration_seconds": self.duration_seconds,
            "view_count": self.view_count,
        }


@dataclass
class YouTubeMetadata:
    """Full YouTube video metadata."""

    video_id: str
    title: str
    description: str
    description_urls: list[str]
    channel_id: str
    channel_title: str
    published_at: str
    duration: str
    duration_seconds: int
    duration_formatted: str
    view_count: int
    like_count: int | None
    comment_count: int | None
    tags: list[str]
    category_id: str | None
    thumbnails: dict
    fetched_at: str

    def to_dict(self) -> dict:
        """Convert to dictionary for JSON serialization."""
        return {
            "video_id": self.video_id,
            "title": self.title,
            "description": self.description,
            "description_urls": self.description_urls,
            "channel_id": self.channel_id,
            "channel_title": self.channel_title,
            "published_at": self.published_at,
            "duration": self.duration,
            "duration_seconds": self.duration_seconds,
            "duration_formatted": self.duration_formatted,
            "view_count": self.view_count,
            "like_count": self.like_count,
            "comment_count": self.comment_count,
            "tags": self.tags,
            "category_id": self.category_id,
            "thumbnails": self.thumbnails,
            "fetched_at": self.fetched_at,
        }


def extract_urls(text: str) -> list[str]:
    """Extract all URLs from text (e.g., video description)."""
    url_pattern = r'https?://[^\s<>"{}|\\^`\[\]]+'
    urls = re.findall(url_pattern, text)

    # Clean trailing punctuation
    cleaned_urls = []
    for url in urls:
        url = url.rstrip(".,;:!?)")
        if url.endswith(")") and url.count("(") < url.count(")"):
            url = url.rstrip(")")
        cleaned_urls.append(url)

    # Deduplicate while preserving order
    seen = set()
    unique_urls = []
    for url in cleaned_urls:
        if url not in seen:
            seen.add(url)
            unique_urls.append(url)

    return unique_urls


def parse_duration_to_seconds(duration: str) -> int:
    """Parse ISO 8601 duration to seconds."""
    match = re.match(r"PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?", duration)
    if not match:
        return 0

    hours, minutes, seconds = match.groups()
    hours = int(hours) if hours else 0
    minutes = int(minutes) if minutes else 0
    seconds = int(seconds) if seconds else 0

    return hours * 3600 + minutes * 60 + seconds


def format_duration(duration: str) -> str:
    """Format ISO 8601 duration to human-readable format."""
    match = re.match(r"PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?", duration)
    if not match:
        return duration

    hours, minutes, seconds = match.groups()
    hours = int(hours) if hours else 0
    minutes = int(minutes) if minutes else 0
    seconds = int(seconds) if seconds else 0

    if hours > 0:
        return f"{hours}:{minutes:02d}:{seconds:02d}"
    return f"{minutes}:{seconds:02d}"


class YouTubeMetadataService:
    """Service for fetching YouTube video metadata using Data API v3."""

    def __init__(self, api_key: str | None = None):
        """Initialize metadata service."""
        self.api_key = api_key or settings.youtube_api_key
        self._youtube = None

    def _get_client(self):
        """Lazy-load the YouTube API client."""
        if self._youtube is None:
            if not self.api_key:
                raise ValueError("YouTube API key not configured")
            from googleapiclient.discovery import build

            self._youtube = build("youtube", "v3", developerKey=self.api_key)
        return self._youtube

    def fetch_metadata(self, video_id: str) -> YouTubeMetadata:
        """Fetch metadata for a YouTube video.

        Args:
            video_id: YouTube video ID (11 characters)

        Returns:
            YouTubeMetadata with all available fields

        Raises:
            ValueError: If video not found or API error
        """
        youtube = self._get_client()

        request = youtube.videos().list(
            part="snippet,statistics,contentDetails",
            id=video_id,
        )
        response = request.execute()

        if not response.get("items"):
            raise ValueError(f"Video not found: {video_id}")

        video = response["items"][0]
        snippet = video["snippet"]
        statistics = video.get("statistics", {})
        content_details = video["contentDetails"]

        duration_iso = content_details["duration"]
        description = snippet.get("description", "")

        return YouTubeMetadata(
            video_id=video_id,
            title=snippet["title"],
            description=description,
            description_urls=extract_urls(description),
            channel_id=snippet["channelId"],
            channel_title=snippet["channelTitle"],
            published_at=snippet["publishedAt"],
            duration=duration_iso,
            duration_seconds=parse_duration_to_seconds(duration_iso),
            duration_formatted=format_duration(duration_iso),
            view_count=int(statistics.get("viewCount", 0)),
            like_count=int(statistics["likeCount"]) if "likeCount" in statistics else None,
            comment_count=int(statistics["commentCount"]) if "commentCount" in statistics else None,
            tags=snippet.get("tags", []),
            category_id=snippet.get("categoryId"),
            thumbnails=snippet.get("thumbnails", {}),
            fetched_at=datetime.now().isoformat(),
        )

    def resolve_channel_id(self, channel: str) -> str:
        """Resolve a YouTube @handle or channel URL to a channel ID."""
        youtube = self._get_client()
        channel = channel.strip().rstrip("/")
        match = re.search(r"youtube\.com/channel/([^/?#]+)", channel)
        if match:
            return match.group(1)
        handle = channel
        if "youtube.com/@" in channel:
            handle = channel.split("youtube.com/@", 1)[1].split("/", 1)[0]
        elif channel.startswith("@"):
            handle = channel[1:]
        else:
            raise ValueError("channel must be an @handle or https://www.youtube.com/@handle")

        response = (
            youtube.search()
            .list(
                part="snippet",
                q=handle,
                type="channel",
                maxResults=1,
            )
            .execute()
        )
        if not response.get("items"):
            raise ValueError(f"No channel found for @{handle}")
        return response["items"][0]["snippet"]["channelId"]

    def fetch_channel_videos(self, channel: str, limit: int = 50) -> list[YouTubeChannelVideo]:
        """Fetch recent uploads for a YouTube @handle or channel URL."""
        youtube = self._get_client()
        channel_id = self.resolve_channel_id(channel)
        channels_response = (
            youtube.channels()
            .list(
                part="contentDetails",
                id=channel_id,
            )
            .execute()
        )
        if not channels_response.get("items"):
            raise ValueError(f"No channel found with ID: {channel_id}")
        playlist_id = channels_response["items"][0]["contentDetails"]["relatedPlaylists"]["uploads"]

        video_ids: list[str] = []
        published: dict[str, str] = {}
        titles: dict[str, str] = {}
        page_token = None
        while len(video_ids) < limit:
            response = (
                youtube.playlistItems()
                .list(
                    part="snippet,contentDetails",
                    playlistId=playlist_id,
                    maxResults=min(50, limit - len(video_ids)),
                    pageToken=page_token,
                )
                .execute()
            )
            for item in response.get("items", []):
                video_id = item["contentDetails"]["videoId"]
                video_ids.append(video_id)
                published[video_id] = item["snippet"]["publishedAt"]
                titles[video_id] = item["snippet"]["title"]
            page_token = response.get("nextPageToken")
            if not page_token:
                break

        videos: list[YouTubeChannelVideo] = []
        for index in range(0, len(video_ids), 50):
            batch = video_ids[index : index + 50]
            details = (
                youtube.videos()
                .list(
                    part="snippet,statistics,contentDetails",
                    id=",".join(batch),
                )
                .execute()
            )
            by_id: dict[str, dict[str, Any]] = {
                item["id"]: item for item in details.get("items", [])
            }
            for video_id in batch:
                item = by_id.get(video_id)
                duration = None
                duration_seconds = None
                view_count = None
                title = titles[video_id]
                if item:
                    title = item["snippet"]["title"]
                    duration_iso = item["contentDetails"]["duration"]
                    duration = format_duration(duration_iso)
                    duration_seconds = parse_duration_to_seconds(duration_iso)
                    stats = item.get("statistics", {})
                    view_count = int(stats["viewCount"]) if "viewCount" in stats else None
                videos.append(
                    YouTubeChannelVideo(
                        video_id=video_id,
                        title=title,
                        url=f"https://www.youtube.com/watch?v={video_id}",
                        published_at=published[video_id],
                        duration=duration,
                        duration_seconds=duration_seconds,
                        view_count=view_count,
                    )
                )
        return videos

    def fetch_metadata_safe(self, video_id: str) -> tuple[YouTubeMetadata | None, str | None]:
        """Fetch metadata with error handling.

        Returns:
            Tuple of (metadata, error_string)
        """
        try:
            metadata = self.fetch_metadata(video_id)
            return metadata, None
        except Exception as e:
            return None, str(e)


def get_youtube_metadata_service() -> YouTubeMetadataService:
    """Get YouTube metadata service instance."""
    return YouTubeMetadataService()

File: services/core/src/vault/youtube-metadata.ts
export type YouTubeChannelVideo = {
  videoId: string;
  title: string;
  url: string;
  publishedAt: string;
  duration: string | null;
  durationSeconds: number | null;
  viewCount: number | null;
};

export type YouTubeMetadata = {
  videoId: string;
  title: string;
  description: string;
  descriptionUrls: string[];
  channelId: string;
  channelTitle: string;
  publishedAt: string;
  duration: string;
  durationSeconds: number;
  durationFormatted: string;
  viewCount: number;
  likeCount: number | null;
  commentCount: number | null;
  tags: string[];
  categoryId: string | null;
  thumbnails: Record<string, unknown>;
  fetchedAt: string;
};

export type YouTubeChannelVideos = {
  source: "youtube-data-api-v3";
  videos: YouTubeChannelVideo[];
};

export type YouTubeMetadataFetcher = (url: string, init?: RequestInit) => Promise<Response>;

const YOUTUBE_DATA_API = "https://www.googleapis.com/youtube/v3";
const YOUTUBE_DATA_API_SOURCE = "youtube-data-api-v3" as const;
const DESCRIPTION_URL = /https?:\/\/[^\s<>"{}|\\^`\[\]]+/g;
const ISO_DURATION = /^PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?/;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function stringAt(record: Record<string, unknown>, key: string, fallback = ""): string {
  const value = record[key];
  return typeof value === "string" ? value : fallback;
}

function integerAt(record: Record<string, unknown>, key: string): number | undefined {
  const value = record[key];
  if (typeof value === "number" && Number.isFinite(value)) return Math.trunc(value);
  if (typeof value !== "string") return undefined;
  const parsed = Number.parseInt(value, 10);
  return Number.isFinite(parsed) ? parsed : undefined;
}

function objectAt(record: Record<string, unknown>, key: string): Record<string, unknown> {
  const value = record[key];
  return isRecord(value) ? value : {};
}

function stringsAt(record: Record<string, unknown>, key: string): string[] {
  const value = record[key];
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : [];
}

function itemsAt(value: unknown): Record<string, unknown>[] {
  return isRecord(value) && Array.isArray(value.items)
    ? value.items.filter((item): item is Record<string, unknown> => isRecord(item))
    : [];
}

function cleanUrl(url: string): string {
  let result = url.replace(/[.,;:!?)]*$/, "");
  if (result.endsWith(")") && (result.match(/\(/g)?.length ?? 0) < (result.match(/\)/g)?.length ?? 0)) {
    result = result.replace(/\)+$/, "");
  }
  return result;
}

export function extractUrls(text: string): string[] {
  const seen = new Set<string>();
  const urls: string[] = [];
  for (const match of text.matchAll(DESCRIPTION_URL)) {
    const url = cleanUrl(match[0]);
    if (!seen.has(url)) {
      seen.add(url);
      urls.push(url);
    }
  }
  return urls;
}

export function parseDurationToSeconds(duration: string): number {
  const match = ISO_DURATION.exec(duration);
  if (match === null) return 0;
  const hours = Number.parseInt(match[1] ?? "0", 10);
  const minutes = Number.parseInt(match[2] ?? "0", 10);
  const seconds = Number.parseInt(match[3] ?? "0", 10);
  return hours * 3600 + minutes * 60 + seconds;
}

export function formatDuration(duration: string): string {
  const match = ISO_DURATION.exec(duration);
  if (match === null) return duration;
  const hours = Number.parseInt(match[1] ?? "0", 10);
  const minutes = Number.parseInt(match[2] ?? "0", 10);
  const seconds = Number.parseInt(match[3] ?? "0", 10);
  if (hours > 0) return `${hours}:${String(minutes).padStart(2, "0")}:${String(seconds).padStart(2, "0")}`;
  return `${minutes}:${String(seconds).padStart(2, "0")}`;
}

function defaultFetcher(url: string, init?: RequestInit): Promise<Response> {
  return fetch(url, init);
}

export class YouTubeMetadataService {
  private readonly fetcher: YouTubeMetadataFetcher;

  constructor(private readonly apiKey: string | undefined, fetcher: YouTubeMetadataFetcher = defaultFetcher) {
    this.fetcher = fetcher;
  }

  private async request(path: string, params: Record<string, string>): Promise<unknown> {
    if (this.apiKey === undefined || this.apiKey === "") throw new Error("YouTube API key not configured");
    const query = new URLSearchParams({ ...params, key: this.apiKey });
    const response = await this.fetcher(`${YOUTUBE_DATA_API}${path}?${query.toString()}`);
    if (!response.ok) throw new Error(`YouTube Data API request failed: ${response.status} ${response.statusText}`);
    return response.json() as Promise<unknown>;
  }

  async fetchMetadata(videoId: string): Promise<YouTubeMetadata> {
    const response = await this.request("/videos", { part: "snippet,statistics,contentDetails", id: videoId });
    const video = itemsAt(response)[0];
    if (video === undefined) throw new Error(`Video not found: ${videoId}`);
    const snippet = objectAt(video, "snippet");
    const statistics = objectAt(video, "statistics");
    const contentDetails = objectAt(video, "contentDetails");
    const duration = stringAt(contentDetails, "duration");
    const description = stringAt(snippet, "description");
    return {
      videoId,
      title: stringAt(snippet, "title"),
      description,
      descriptionUrls: extractUrls(description),
      channelId: stringAt(snippet, "channelId"),
      channelTitle: stringAt(snippet, "channelTitle"),
      publishedAt: stringAt(snippet, "publishedAt"),
      duration,
      durationSeconds: parseDurationToSeconds(duration),
      durationFormatted: formatDuration(duration),
      viewCount: integerAt(statistics, "viewCount") ?? 0,
      likeCount: integerAt(statistics, "likeCount") ?? null,
      commentCount: integerAt(statistics, "commentCount") ?? null,
      tags: stringsAt(snippet, "tags"),
      categoryId: stringAt(snippet, "categoryId") || null,
      thumbnails: objectAt(snippet, "thumbnails"),
      fetchedAt: new Date().toISOString(),
    };
  }

  async resolveChannelId(channel: string): Promise<string> {
    const normalized = channel.trim().replace(/\/$/, "");
    const channelMatch = /youtube\.com\/channel\/([^/?#]+)/.exec(normalized);
    if (channelMatch?.[1] !== undefined) return channelMatch[1];

    let handle: string;
    const urlHandle = /youtube\.com\/@([^/?#]+)/.exec(normalized);
    if (urlHandle?.[1] !== undefined) {
      handle = urlHandle[1];
    } else if (normalized.startsWith("@")) {
      handle = normalized.slice(1);
    } else {
      throw new Error("channel must be an @handle or https://www.youtube.com/@handle");
    }

    const response = await this.request("/search", { part: "snippet", q: handle, type: "channel", maxResults: "1" });
    const channelItem = itemsAt(response)[0];
    const channelId = channelItem === undefined ? "" : stringAt(objectAt(channelItem, "snippet"), "channelId");
    if (channelId === "") throw new Error(`No channel found for @${handle}`);
    return channelId;
  }

  async fetchChannelVideos(channel: string, limit = 50): Promise<YouTubeChannelVideo[]> {
    const channelId = await this.resolveChannelId(channel);
    const channels = await this.request("/channels", { part: "contentDetails", id: channelId });
    const channelItem = itemsAt(channels)[0];
    const uploads = channelItem === undefined
      ? ""
      : stringAt(objectAt(objectAt(channelItem, "contentDetails"), "relatedPlaylists"), "uploads");
    if (uploads === "") throw new Error(`No channel found with ID: ${channelId}`);

    const videoIds: string[] = [];
    const published = new Map<string, string>();
    const titles = new Map<string, string>();
    let pageToken: string | undefined;
    while (videoIds.length < limit) {
      const params: Record<string, string> = {
        part: "snippet,contentDetails",
        playlistId: uploads,
        maxResults: String(Math.min(50, limit - videoIds.length)),
      };
      if (pageToken !== undefined) params.pageToken = pageToken;
      const page = await this.request("/playlistItems", params);
      for (const item of itemsAt(page)) {
        const videoId = stringAt(objectAt(item, "contentDetails"), "videoId");
        if (videoId === "") continue;
        const snippet = objectAt(item, "snippet");
        videoIds.push(videoId);
        published.set(videoId, stringAt(snippet, "publishedAt"));
        titles.set(videoId, stringAt(snippet, "title"));
      }
      pageToken = isRecord(page) ? stringAt(page, "nextPageToken") || undefined : undefined;
      if (pageToken === undefined) break;
    }

    const details = new Map<string, Record<string, unknown>>();
    for (let index = 0; index < videoIds.length; index += 50) {
      const batch = videoIds.slice(index, index + 50);
      const response = await this.request("/videos", { part: "snippet,statistics,contentDetails", id: batch.join(",") });
      for (const item of itemsAt(response)) {
        const id = stringAt(item, "id");
        if (id !== "") details.set(id, item);
      }
    }

    return videoIds.map((videoId) => {
      const item = details.get(videoId);
      if (item === undefined) {
        return {
          videoId,
          title: titles.get(videoId) ?? "",
          url: `https://www.youtube.com/watch?v=${videoId}`,
          publishedAt: published.get(videoId) ?? "",
          duration: null,
          durationSeconds: null,
          viewCount: null,
        };
      }
      const duration = stringAt(objectAt(item, "contentDetails"), "duration");
      return {
        videoId,
        title: stringAt(objectAt(item, "snippet"), "title", titles.get(videoId) ?? ""),
        url: `https://www.youtube.com/watch?v=${videoId}`,
        publishedAt: published.get(videoId) ?? "",
        duration: formatDuration(duration),
        durationSeconds: parseDurationToSeconds(duration),
        viewCount: integerAt(objectAt(item, "statistics"), "viewCount") ?? null,
      };
    });
  }

  async fetchChannelVideosResponse(channel: string, limit = 50): Promise<YouTubeChannelVideos> {
    return { source: YOUTUBE_DATA_API_SOURCE, videos: await this.fetchChannelVideos(channel, limit) };
  }

  async fetchMetadataSafe(videoId: string): Promise<{ metadata: YouTubeMetadata | null; error: string | null }> {
    try {
      return { metadata: await this.fetchMetadata(videoId), error: null };
    } catch (error) {
      return { metadata: null, error: error instanceof Error ? error.message : String(error) };
    }
  }
}
